asllib/doc/Expressions.tex

```diff
@@ -1691,45 +1779,286 @@ \subsubsection{Formally}
+      \item applying $\widthplus$ to the list of expressions $\ewidth_\vf$, for each $\vf$ in $\tfields$, yields $\vw$\ProseOrTypeError;
+      \item define $\vt$ as the bitvector type with width $\vw$ and an empty list of bitfields;
+      \item define $\ve$ as the multi-field access for $\vetwo$ and list of fields $\tfields$;
+      \item define $\vses$ as $\vsesone$.
+    \end{itemize}
+  \end{itemize}
+\end{itemize}
 
 \subsubsection{Formally}
 \begin{mathpar}
-  \inferrule[empty]{}
-  {
-    \evalexprlist{\env, \emptylist} \evalarrow \Normal((\emptylist, \emptygraph), \env)
-  }
-\and
-  \inferrule[nonempty]{
-    \texttt{le} \eqname [\ve] \concat \vle1\\
-    \evalexpr{\env, \ve} \evalarrow \Normal((\vvone, \vgone), \envone) \OrAbnormal\\
-    \evalexprlist{\envone, \vle1} \evalarrow \Normal((\vvs, \vgtwo), \newenv) \OrAbnormal\\
-    \vg \eqdef \vgone \parallelcomp \vgtwo \\
-    \vv \eqdef [\vvone] \concat \vvs
-  }
+\inferrule[bits]{
+  \annotateexpr{\tenv, \veone} \typearrow (\vtetwo, \vetwo, \vsesone) \OrTypeError\\\\
+  \makeanonymous(\tenv, \veone) \typearrow \TBits(\Ignore, \vbitfields) \OrTypeError\\\\
+  \name\in\vfields: \findbitfieldsslices(\name) \typearrow \vslices_\name \OrTypeError\\\\
+  \vep \eqdef \ESlice(\veone, [\name\in\vfields: \vslices_\name])\\
+  \annotateexpr{\tenv, \vep} \typearrow (\vt, \newe, \vses) \OrTypeError
+}{
+  \annotateexpr{\tenv, \overname{\EGetFields(\veone, \vfields)}{\ve}} \typearrow (\vt, \newe, \vses)
+}
+\end{mathpar}
+
+\begin{mathpar}
+\inferrule[record]{
+  \annotateexpr{\tenv, \veone} \typearrow (\vtetwo, \vetwo, \vsesone) \OrTypeError\\\\
+  \makeanonymous(\tenv, \veone) \typearrow \TRecord(\tfields) \OrTypeError\\\\
+  \vf \in \tfields: \getbitfieldwidth(\tenv, \vf, \tfields) \typearrow \ewidth_\vf \OrTypeError\\\\
+  \widthplus(\tenv, [\vf \in \tfields: \ewidth_\vf]) \typearrow \vw\OrTypeError
+}{
   {
-    \evalexprlist{\env, \vle} \evalarrow \Normal((\vv, \vg), \newenv)
+  \begin{array}{r}
+    \annotateexpr{\tenv, \overname{\EGetFields(\veone, \vfields)}{\ve}} \typearrow \\
+    (\overname{\TBits(\vw, \emptylist)}{\vt}, \overname{\EGetFields(\vetwo, \vfields)}{\newe}, \overname{\vsesone}{\vses})
+  \end{array}
   }
+}
 \end{mathpar}
 
+\begin{mathpar}
+\inferrule[error]{
+  \annotateexpr{\tenv, \veone} \typearrow (\vtetwo, \vetwo, \vsesone) \OrTypeError\\\\
+  \makeanonymous(\tenv, \veone) \typearrow \vtetwoanon \OrTypeError\\\\
+  \astlabel(\vtetwoanon) \not\in \{\TBits, \TRecord\}
+}{
+  \annotateexpr{\tenv, \overname{\EGetFields(\veone, \vfields)}{\ve}} \typearrow \TypeErrorVal{}
+}
+\end{mathpar}
+\CodeSubsection{\EGetFieldsBegin}{\EGetFieldsEnd}{../Typing.ml}
+
+\subsubsection{TypingRule.FindBitFieldsSlices}
+\hypertarget{def-findbitfieldsslices}{}
+% NOTE: The implementation returns an optional, but all of its uses raise a type error if the result is None.
```

Comment:
is this still true?

Reply:
Yes, this is just me documenting slight digressions from the implementation.